### ISSUE TITLE
compile issue in pin.cpp

### ISSUE THREAD

**Issue by Sir-Ace-git** [dimitry-ishenko-cpp/gpio#5]
I've got a compile issue in pin.cpp 
My asio version, kernel version/platform, cmake version, and gcc version are below, as well as a list of all the headers with references to the sleep until function.  The board is a RockPro64 but that shouldn't matter for the compile.

root@asteroid:~/gpio/build# cmake .. -DCMAKE_INSTALL_PREFIX=/usr
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /root/gpio/build

root@asteroid:~/gpio/build# make
[ 12%] Building CXX object base/CMakeFiles/gpio++-base.dir/chip_base.cpp.o
[ 25%] Building CXX object base/CMakeFiles/gpio++-base.dir/pin_base.cpp.o
[ 25%] Built target gpio++-base
[ 37%] Building CXX object generic/CMakeFiles/gpio++.dir/chip.cpp.o
[ 50%] Building CXX object generic/CMakeFiles/gpio++.dir/pin.cpp.o
/root/gpio/generic/pin.cpp: In lambda function:
/root/gpio/generic/pin.cpp:275:31: error: ‘sleep_until’ is not a member of ‘std::this_thread’
  275 |             std::this_thread::sleep_until(tp);
      |                               ^~~~~~~~~~~
/root/gpio/generic/pin.cpp:280:31: error: ‘sleep_until’ is not a member of ‘std::this_thread’
  280 |             std::this_thread::sleep_until(tp);
      |                               ^~~~~~~~~~~
make[2]: *** [generic/CMakeFiles/gpio++.dir/build.make:90: generic/CMakeFiles/gpio++.dir/pin.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:160: generic/CMakeFiles/gpio++.dir/all] Error 2
make: *** [Makefile:156: all] Error 2

Using:
asio-1.20.0

root@asteroid:~/gpio/build# uname -a
Linux asteroid.wizardsworks.org 5.15.19-armv8 #1 SMP PREEMPT Wed Feb 2 08:52:45 GMT 2022 aarch64 GNU/Linux

root@asteroid:~/gpio/build# gcc -v
Reading specs from /usr/lib64/gcc/aarch64-slackware-linux-gnu/11.2.0/specs
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/aarch64-slackware-linux-gnu/11.2.0/lto-wrapper
Target: aarch64-slackware-linux-gnu
Configured with: ../configure --with-arch=armv8-a --verbose --prefix=/usr --mandir=/usr/man --infodir=/usr/info --libdir=/usr/lib64 --enable-bootstrap --enable-checking=release --enable-libstdcxx-dual-abi --enable-shared --enable-languages=ada,c,c++,d,fortran,go,lto,objc,obj-c++ --enable-objc-gc --enable-threads=posix --enable-__cxa_atexit --enable-gnu-unique-object --enable-clocale=gnu --enable-plugin --enable-lto --with-arch-directory= --with-system-zlib --with-gnu-ld --with-isl --with-default-libstdcxx-abi=new --disable-libunwind-exceptions --disable-libstdcxx-pch --disable-libssp --disable-werror --disable-gtktest --disable-install-libiberty --host=aarch64-slackware-linux-gnu --build=aarch64-slackware-linux-gnu --target=aarch64-slackware-linux-gnu
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (GCC)

root@asteroid:~/gpio/build# cmake --version
cmake version 3.21.4

CMake suite maintained and supported by Kitware (kitware.com/cmake).

Here is what I found for the function in my headers:
root@asteroid:~/gpio/build# grep sleep_until -R /usr/include/
/usr/include/boost/thread/win32/thread_data.hpp:        void sleep_until(const chrono::time_point<chrono::steady_clock, Duration>& t)
/usr/include/boost/thread/win32/thread_data.hpp:        void sleep_until(const chrono::time_point<Clock, Duration>& t)
/usr/include/boost/thread/win32/thread_data.hpp:          void sleep_until(const chrono::time_point<chrono::steady_clock, Duration>& t)
/usr/include/boost/thread/win32/thread_data.hpp:          void sleep_until(const chrono::time_point<Clock, Duration>& t)
/usr/include/boost/thread/pthread/thread_data.hpp:        void sleep_until(const chrono::time_point<Clock, Duration>& t)
/usr/include/boost/thread/pthread/thread_data.hpp:          void sleep_until(const chrono::time_point<chrono::steady_clock, Duration>& t)
/usr/include/boost/thread/pthread/thread_data.hpp:          void sleep_until(const chrono::time_point<Clock, Duration>& t)
/usr/include/boost/thread/pthread/thread_data.hpp:          void sleep_until(const chrono::time_point<Clock, Duration>& t)
/usr/include/boost/thread/pthread/thread_data.hpp:            this_thread::sleep_until(t);
/usr/include/boost/asio/detail/impl/null_event.ipp:  std::this_thread::sleep_until((std::chrono::steady_clock::time_point::max)());
/usr/include/boost/fiber/operations.hpp:void sleep_until( std::chrono::time_point< Clock, Duration > const& sleep_time_) {
/usr/include/php/ext/standard/basic_functions.h:PHP_FUNCTION(time_sleep_until);
/usr/include/c++/11.2.0/thread:#include <bits/this_thread_sleep.h> // std::this_thread::sleep_for, sleep_until
/usr/include/c++/11.2.0/bits/this_thread_sleep.h:    /// this_thread::sleep_until
/usr/include/c++/11.2.0/bits/this_thread_sleep.h:      sleep_until(const chrono::time_point<_Clock, _Duration>& __atime)
/usr/include/c++/v1/thread:void sleep_until(const chrono::time_point<Clock, Duration>& abs_time);
/usr/include/c++/v1/thread:sleep_until(const chrono::time_point<_Clock, _Duration>& __t)
/usr/include/c++/v1/thread:sleep_until(const chrono::time_point<chrono::steady_clock, _Duration>& __t)

**Comment by dimitry-ishenko** [comment]
@Sir-Ace-git it seems that maybe your platform does not implement `std::high_resolution_clock`.

Can you try changing line 271 from this:
```c++
for(auto tp = std::chrono::high_resolution_clock::now();;)
```
to this:
```c++
for(auto tp = std::chrono::steady_clock::now();;)
```
and see if it handles the problem. Thanks.

**Comment by Sir-Ace-git** [comment]
Same problem....  The lines with the error were 275 and 280, which were after the for line you had me change. 
The lines below are 276/281 because I commented the old for line out, and inserted the new one.

With the change:
root@asteroid:~/gpio++/build# make
[ 12%] Building CXX object base/CMakeFiles/gpio++-base.dir/chip_base.cpp.o
[ 25%] Building CXX object base/CMakeFiles/gpio++-base.dir/pin_base.cpp.o
[ 25%] Built target gpio++-base
[ 37%] Building CXX object generic/CMakeFiles/gpio++.dir/chip.cpp.o
[ 50%] Building CXX object generic/CMakeFiles/gpio++.dir/pin.cpp.o
/root/gpio++/generic/pin.cpp: In lambda function:
/root/gpio++/generic/pin.cpp:276:31: error: ‘sleep_until’ is not a member of ‘std::this_thread’
  276 |             std::this_thread::sleep_until(tp);
      |                               ^~~~~~~~~~~
/root/gpio++/generic/pin.cpp:281:31: error: ‘sleep_until’ is not a member of ‘std::this_thread’
  281 |             std::this_thread::sleep_until(tp);
      |                               ^~~~~~~~~~~
make[2]: *** [generic/CMakeFiles/gpio++.dir/build.make:90: generic/CMakeFiles/gpio++.dir/pin.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:160: generic/CMakeFiles/gpio++.dir/all] Error 2
make: *** [Makefile:156: all] Error 2

**Comment by dimitry-ishenko** [comment]
If you don't mind me asking, which box is this?

**Comment by dimitry-ishenko** [comment]
When you run CMake, try adding `_GLIBCXX_USE_NANOSLEEP` define. In other words:
```sh
cmake .. -DCMAKE_INSTALL_PREFIX=/usr -D_GLIBCXX_USE_NANOSLEEP
```

If that doesn't work, try poking around in `/usr/include/c++/v1/thread` and see if it's hidden behind another define.

If you get stuck, attach the file here and I will try to take a look.

**Comment by Sir-Ace-git** [comment]
***@***.***:~/gpio++/build# cmake .. -DCMAKE_INSTALL_PREFIX=/usr 
-D_GLIBCXX_USE_NANOSLEEP
CMake Error: Parse error in command line argument: 
_GLIBCXX_USE_NANOSLEEP
  Should be: VAR:type=value

CMake Error: Run 'cmake --help' for all supported options.



To answer your other question this is Slackware 15 on a RockPro64

**Comment by dimitry-ishenko** [comment]
@Sir-Ace-git sorry that switch is supposed to be used with `make` and not `cmake`. So try this instead:
```sh
cmake .. -DCMAKE_INSTALL_PREFIX=/usr
make -D_GLIBCXX_USE_NANOSLEEP
```

If that doesn't work, try adding `add_definitions(-D_GLIBCXX_USE_NANOSLEEP)` somewhere 
at the top of `CMakeLists.txt`.

**Comment by Sir-Ace-git** [comment]
I started with a brand new git clone on this pass.
make doesn't support the -D flag, so I added it to the CMakeLists.txt 
like you suggested, and that didn't seem to change anything, so I then 
tried adding it to all of the CMakeLists files, which also didn't seem 
to have any effect:

***@***.***:~/gpio++# find ../ |grep CMakeLists.txt
../gpio++/pigpio/CMakeLists.txt
../gpio++/base/CMakeLists.txt
../gpio++/generic/CMakeLists.txt
../gpio++/CMakeLists.txt

***@***.***:~/gpio++# for x in `find ../ |grep CMakeLists.txt` ; do vi 
$x ; done

***@***.***:~/gpio++# for x in `find ../ |grep CMakeLists.txt` ; do 
head -n 5 $x ; done
cmake_minimum_required(VERSION 3.1)
project(gpio++-pigpio VERSION 4.1)

add_definitions(-D_GLIBCXX_USE_NANOSLEEP)

cmake_minimum_required(VERSION 3.1)
project(gpio++-base VERSION 4.1)

add_definitions(-D_GLIBCXX_USE_NANOSLEEP)

cmake_minimum_required(VERSION 3.1)
project(gpio++ VERSION 4.1)

add_definitions(-D_GLIBCXX_USE_NANOSLEEP)

cmake_minimum_required(VERSION 3.1)
project(gpio++ VERSION 4.1)

add_definitions(-D_GLIBCXX_USE_NANOSLEEP)


***@***.***:~/gpio++# mkdir build

***@***.***:~/gpio++# cd build/

***@***.***:~/gpio++/build# cmake .. -DCMAKE_INSTALL_PREFIX=/usr
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /root/gpio++/build

***@***.***:~/gpio++/build# make
[ 12%] Building CXX object 
base/CMakeFiles/gpio++-base.dir/chip_base.cpp.o
[ 25%] Building CXX object 
base/CMakeFiles/gpio++-base.dir/pin_base.cpp.o
[ 25%] Built target gpio++-base
[ 37%] Building CXX object generic/CMakeFiles/gpio++.dir/chip.cpp.o
[ 50%] Building CXX object generic/CMakeFiles/gpio++.dir/pin.cpp.o
/root/gpio++/generic/pin.cpp: In lambda function:
/root/gpio++/generic/pin.cpp:275:31: error: ‘sleep_until’ is not a 
member of ‘std::this_thread’
   275 |             std::this_thread::sleep_until(tp);
       |                               ^~~~~~~~~~~
/root/gpio++/generic/pin.cpp:280:31: error: ‘sleep_until’ is not a 
member of ‘std::this_thread’
   280 |             std::this_thread::sleep_until(tp);
       |                               ^~~~~~~~~~~
make[2]: *** [generic/CMakeFiles/gpio++.dir/build.make:90: 
generic/CMakeFiles/gpio++.dir/pin.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:160: 
generic/CMakeFiles/gpio++.dir/all] Error 2
make: *** [Makefile:156: all] Error 2

**Comment by dimitry-ishenko** [comment]
I've grabbed [gcc-g++-11.2.0-arm-1.txz](http://ftp.halifax.rwth-aachen.de/slackwarearm/slackwarearm-current/slackware/d/gcc-g++-11.2.0-arm-1.txz) from Slackware FTP and looked inside `/usr/include/c++/11.2.0/bits/this_thread_sleep.h`.

This is what I see:
```c++
// std::this_thread::sleep_for/until declarations -*- C++ -*-

// Copyright (C) 2008-2021 Free Software Foundation, Inc.
//
// This file is part of the GNU ISO C++ Library.  This library is free
// software; you can redistribute it and/or modify it under the
// terms of the GNU General Public License as published by the
// Free Software Foundation; either version 3, or (at your option)
// any later version.

// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// Under Section 7 of GPL version 3, you are granted additional
// permissions described in the GCC Runtime Library Exception, version
// 3.1, as published by the Free Software Foundation.

// You should have received a copy of the GNU General Public License and
// a copy of the GCC Runtime Library Exception along with this program;
// see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
// <http://www.gnu.org/licenses/>.

/** @file bits/std_thread_sleep.h
 *  This is an internal header file, included by other library headers.
 *  Do not attempt to use it directly. @headername{thread}
 */

#ifndef _GLIBCXX_THIS_THREAD_SLEEP_H
#define _GLIBCXX_THIS_THREAD_SLEEP_H 1

#pragma GCC system_header

#if __cplusplus >= 201103L
#include <bits/c++config.h>

#include <chrono> // std::chrono::*

#ifdef _GLIBCXX_USE_NANOSLEEP
# include <cerrno>  // errno, EINTR
# include <time.h>  // nanosleep
#endif

namespace std _GLIBCXX_VISIBILITY(default)
{
_GLIBCXX_BEGIN_NAMESPACE_VERSION

  /** @addtogroup threads
   *  @{
   */

  /** @namespace std::this_thread
   *  @brief ISO C++ 2011 namespace for interacting with the current thread
   *
   *  C++11 30.3.2 [thread.thread.this] Namespace this_thread.
   */
  namespace this_thread
  {
#ifndef _GLIBCXX_NO_SLEEP

#ifndef _GLIBCXX_USE_NANOSLEEP
    void
    __sleep_for(chrono::seconds, chrono::nanoseconds);
#endif

    /// this_thread::sleep_for
    template<typename _Rep, typename _Period>
      inline void
      sleep_for(const chrono::duration<_Rep, _Period>& __rtime)
      {
	if (__rtime <= __rtime.zero())
	  return;
	auto __s = chrono::duration_cast<chrono::seconds>(__rtime);
	auto __ns = chrono::duration_cast<chrono::nanoseconds>(__rtime - __s);
#ifdef _GLIBCXX_USE_NANOSLEEP
	struct ::timespec __ts =
	  {
	    static_cast<std::time_t>(__s.count()),
	    static_cast<long>(__ns.count())
	  };
	while (::nanosleep(&__ts, &__ts) == -1 && errno == EINTR)
	  { }
#else
	__sleep_for(__s, __ns);
#endif
      }

    /// this_thread::sleep_until
    template<typename _Clock, typename _Duration>
      inline void
      sleep_until(const chrono::time_point<_Clock, _Duration>& __atime)
      {
#if __cplusplus > 201703L
	static_assert(chrono::is_clock_v<_Clock>);
#endif
	auto __now = _Clock::now();
	if (_Clock::is_steady)
	  {
	    if (__now < __atime)
	      sleep_for(__atime - __now);
	    return;
	  }
	while (__now < __atime)
	  {
	    sleep_for(__atime - __now);
	    __now = _Clock::now();
	  }
      }
  } // namespace this_thread
#endif // ! NO_SLEEP

  /// @}

_GLIBCXX_END_NAMESPACE_VERSION
} // namespace
#endif // C++11

#endif // _GLIBCXX_THIS_THREAD_SLEEP_H
```

**Comment by dimitry-ishenko** [comment]
The file is basically identical to the one on my system. The only thing I can see
is that `std::this_thread::sleep_for` and `std::this_thread::sleep_until` are hidden
behind `_GLIBCXX_NO_SLEEP` macro.

Could you try adding the following at the top of `generic/pin.cpp` just after the includes:
```c++
#ifdef _GLIBCXX_NO_SLEEP
#  error Mayday, mayday, no sleep function detected!!!
#endif
```
and see if you get that error.

Also, maybe try compiling this test and see if you get an error as well:
```c++
#include <thread>
int main()
{
    std::this_thread::sleep_for(std::chrono::seconds(3));
    return 0;
}
```
This should tell you if there is some mis-configuration in your system, or maybe CMake is doing something strange.

**Comment by dimitry-ishenko** [comment]
Another thing I've noticed is that my compiler includes `--enable-libstdcxx-time` option, where as yours doesn't:
```sh
dimitry@laptop:~$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/11/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 11.2.0-7ubuntu2' 
--with-bugurl=file:///usr/share/doc/gcc-11/README.Bugs 
--enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only 
--program-suffix=-11 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id 
--libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls 
--enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes 
--with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin 
--enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto 
--enable-objc-gc=auto --enable-multiarch --disable-werror --enable-cet --with-arch-32=i686 
--with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic 
--enable-offload-targets=nvptx-none=/build/gcc-11-ZPT0kp/gcc-11-11.2.0/debian/tmp-nvptx/usr,amdgcn-amdhsa=/build/gcc-11-ZPT0kp/gcc-11-11.2.0/debian/tmp-gcn/usr 
--without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu 
--target=x86_64-linux-gnu --with-build-config=bootstrap-lto-lean --enable-link-serialization=2
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Ubuntu 11.2.0-7ubuntu2) 
```

If the 2nd test in the previous comment gives you an error, that means your compiler and/or standard library was compiled without support for sleep functions.

**Comment by Sir-Ace-git** [comment]
I think you might have gotten it with this one:
***@***.***:/var/log/packages# gcc -v 2>&1 |tr " " "\n" |grep 
enable-libstdcxx
--enable-libstdcxx-dual-abi

It does not appear to support the sleep functions.

**Comment by Sir-Ace-git** [comment]
***@***.***:/tmp# cat 1.cpp
#include <thread>
int main()
{
     std::this_thread::sleep_for(std::chrono::seconds(3));
     return 0;
}

***@***.***:/tmp# gcc 1.cpp

***@***.***:/tmp# time ./a.out

real    0m3.007s
user    0m0.002s
sys     0m0.005s



Also, I did add the ifdef section, and it did not change anything:

[ 37%] Building CXX object generic/CMakeFiles/gpio++.dir/pin.cpp.o
/root/gpio++/generic/pin.cpp: In lambda function:
/root/gpio++/generic/pin.cpp:281:31: error: ‘sleep_until’ is not a 
member of ‘std::this_thread’
   281 |             std::this_thread::sleep_until(tp);
       |                               ^~~~~~~~~~~
/root/gpio++/generic/pin.cpp:286:31: error: ‘sleep_until’ is not a 
member of ‘std::this_thread’
   286 |             std::this_thread::sleep_until(tp);
       |                               ^~~~~~~~~~~
make[2]: *** [generic/CMakeFiles/gpio++.dir/build.make:90: 
generic/CMakeFiles/gpio++.dir/pin.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:160: 
generic/CMakeFiles/gpio++.dir/all] Error 2
make: *** [Makefile:156: all] Error 2

**Comment by dimitry-ishenko** [comment]
Can you run `make V=1` and post the output?

Also, could you attach a link to the image of Slackware you are using?
I'll see if I can reproduce it on my end.

**Comment by Sir-Ace-git** [comment]
Here is the make output:
***@***.***:~/gpio++/build# make clean

***@***.***:~/gpio++/build# make V=1
[ 12%] Building CXX object 
base/CMakeFiles/gpio++-base.dir/chip_base.cpp.o
[ 25%] Building CXX object 
base/CMakeFiles/gpio++-base.dir/pin_base.cpp.o
[ 25%] Built target gpio++-base
[ 37%] Building CXX object generic/CMakeFiles/gpio++.dir/chip.cpp.o
[ 50%] Building CXX object generic/CMakeFiles/gpio++.dir/pin.cpp.o
/root/gpio++/generic/pin.cpp: In lambda function:
/root/gpio++/generic/pin.cpp:281:31: error: ‘sleep_until’ is not a 
member of ‘std::this_thread’
   281 |             std::this_thread::sleep_until(tp);
       |                               ^~~~~~~~~~~
/root/gpio++/generic/pin.cpp:286:31: error: ‘sleep_until’ is not a 
member of ‘std::this_thread’
   286 |             std::this_thread::sleep_until(tp);
       |                               ^~~~~~~~~~~
make[2]: *** [generic/CMakeFiles/gpio++.dir/build.make:90: 
generic/CMakeFiles/gpio++.dir/pin.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:160: 
generic/CMakeFiles/gpio++.dir/all] Error 2
make: *** [Makefile:156: all] Error 2



The Slackware 15 ARM64 port has not been released yet (I am helping test 
it out), so I've been asked not to share it yet.
I can however make the machine available for login to you if you would 
like that.  Contact me directly via email for login info for the 
machine.

**Comment by dimitry-ishenko** [comment]
@Sir-Ace-git this was fixed in a8d74a9755676a2bbe22e9d5916fbdd57eed888b. Thanks for your patience.
